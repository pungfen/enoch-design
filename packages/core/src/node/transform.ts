import { compileScript, compileTemplate, parse } from 'vue/compiler-sfc'
import MagicStringBase from 'magic-string'

import type { SFCParseResult } from 'vue/compiler-sfc'
import type { Options } from './plugin'

const DEFINE_FACTORY = 'defineFactory'

class MagicCode extends MagicStringBase {}

const transformCustomBlock = (id: string, sfc: SFCParseResult, options: Options, ms: MagicCode) => {
  if (!sfc.descriptor.customBlocks) return
  const customBlocks = sfc.descriptor.customBlocks
  for (const block of customBlocks) {
    if (block.type === 'pung' && block.lang === 'json') {
      // const blocConfig = JSON.parse(block.content)
      // ms.setConfig(blocConfig)
    }
  }
}

const transformTemplate = (id: string, sfc: SFCParseResult, options: Options, ms: MagicCode) => {
  if (!sfc.descriptor.template) return
}

const transformScript = (id: string, sfc: SFCParseResult, options: Options, ms: MagicCode) => {
  if (sfc.descriptor.scriptSetup) {
    const compiled = compileScript(sfc.descriptor, { id })
    const start = compiled.loc.start
    const end = compiled.loc.end
    let cfg: string = ''
    for (const node of compiled.scriptSetupAst!) {
      switch (node.type) {
        case 'ExpressionStatement':
          const expression = node.expression
          if (expression.type === 'CallExpression' && expression.callee.type === 'Identifier' && expression.callee.name === DEFINE_FACTORY) {
            const [_config, _options] = expression.arguments
            if (_config.type === 'ObjectExpression' && _config.properties) {
              cfg = ms.snip(start.offset + _config.start!, start.offset + _config.end!).toString()
            }
            ms.remove(node.start! + start.offset, node.end! + start.offset)
          }
          break
      }
    }
    if (cfg) {
      ms.prependLeft(
        start.offset,
        `
        \nimport { block as _DF_block, proxy as _DF_proxy } from '@enochfe/core';
        \nconst _DF_BLOCK = _DF_block(${cfg});
        \nconst _DF_PROXY = _DF_proxy(_DF_BLOCK);
      `
      )
    }
  }
}

export const transform = (code: string, id: string, options: Options) => {
  if (!code.includes(DEFINE_FACTORY)) return

  const sfc = parse(code, { filename: id })
  const ms = new MagicCode(code, { filename: id })

  transformCustomBlock(id, sfc, options, ms)
  transformScript(id, sfc, options, ms)
  transformTemplate(id, sfc, options, ms)

  return (
    (ms.hasChanged() && {
      code: ms.toString(),
      get map() {
        return ms.generateMap({
          source: id,
          includeContent: true,
          hires: true
        })
      }
    }) ||
    undefined
  )
}
