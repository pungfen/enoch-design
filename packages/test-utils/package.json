{
  "name": "@enochfe/test-utils",
  "author": "pung <fp94214@gmail.com>",
  "description": "enoch test-utils",
  "version": "1.0.1",
  "type": "module",
  "main": "src/index.ts",
  "scripts": {
    "build": "tsup"
  },
  "dependencies": {
    "@vitejs/plugin-vue-jsx": "^2.0.1",
    "esbuild": "^0.15.10",
    "rollup": "^3.1.0",
    "rollup-plugin-esbuild": "^4.10.1",
    "unplugin": "^0.9.6",
    "unplugin-vue": "3.1.3"
  },
  "devDependencies": {
    "vite": "^3.1.3",
    "vue": "^3.2.39"
  }
}
