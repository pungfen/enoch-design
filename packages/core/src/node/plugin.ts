import { bold, green } from 'kolorist'
import { createUnplugin } from 'unplugin'

import { transform } from './transform'

import type { UnpluginFactory } from 'unplugin'
import type { ResolvedConfig, ViteDevServer } from 'vite'

export interface Options {
  platform?: 'vue' | 'react' | 'weapp'
}

const name = 'unplugin-enochfe'

const unplugin: UnpluginFactory<Options> = (options: Options = {}) => {
  let config: ResolvedConfig

  return {
    name,
    enforce: 'pre',
    vite: {
      configResolved(_config) {
        config = _config
      },
      configureServer(server: ViteDevServer) {
        // const _print = server.printUrls
        // server.printUrls = () => {
        //   const colorUrl = (url: string) => green(url.replace(/:(\d+)\//, (_, port) => `:${bold(port)}/`))
        //   const host = server.resolvedUrls?.local[0] || `${config.server.https ? 'https' : 'http'}://localhost:${config.server.port || '80'}`
        //   _print()
        //   console.log(`  ${green('âžœ')}  ${bold('Inspect')}: ${colorUrl(`${host}__core/`)}`)
        // }
      }
    },

    transform(code, id) {
      try {
        return transform(code, id, options)
      } catch (err: unknown) {
        this.error(`${name} ${err}`)
      }
    }
  }
}

export default createUnplugin(unplugin)
