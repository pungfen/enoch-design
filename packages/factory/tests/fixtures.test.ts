import { resolve } from 'node:path'
import glob from 'fast-glob'

import { RollupToStringPlugin, rollupBuild } from '@enochfe/test-utils'
import Factory from '../src/rollup'

describe('Rollup', () => {
  describe('fixtures', async () => {
    const root = resolve(__dirname, '..')

    const files = await glob('tests/fixtures/*.vue', { cwd: root, onlyFiles: true })

    for (const file of files) {
      it(file.replace(/\\/g, '/'), async () => {
        const filepath = resolve(root, file)

        const code = await rollupBuild(filepath, [Factory(), RollupToStringPlugin()]).catch((err) => err)

        expect(code).toMatchSnapshot()
      })
    }
  })
})
