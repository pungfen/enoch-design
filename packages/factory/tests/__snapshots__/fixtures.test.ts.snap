// Vitest Snapshot v1

exports[`Rollup > fixtures > tests/fixtures/button.vue 1`] = `
"var button = \`<script lang=\\"ts\\">
import { defineComponent as _defineComponent } from 'vue';
export default /*#__PURE__*/ _defineComponent(
{
  name: 'en-button',
  render() {
    return h('div', this.$props)
  }
}
);

</script>
<script setup lang=\\"ts\\">
import { h } from 'vue'


</script>

<template>
  <div>
    <h1>hello world</h1>
  </div>
</template>
\`;

export { button as default };
"
`;
